File: BOJ/Kotlin/그리디 알고리즘/1202.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*
import kotlin.collections.ArrayList

data class Jewel(val m: Int, val v: Int)

fun main() = with(BufferedReader(InputStreamReader(System.`in`))){
    val (N, K) = readLine().split(" ").map { it.toInt() }
    var res = 0L
    val priorityQueue = PriorityQueue<Int>(Collections.reverseOrder())
    val jewelArr = ArrayList<Jewel>()
    val bagArr = ArrayList<Int>()

    for (i in 0 until N){
        val (m, v) = readLine().split(" ").map { it.toInt() }
        jewelArr.add(Jewel(m, v))
    }

    for (i in 0 until K)
        bagArr.add(readLine().toInt())

    bagArr.sort()
    jewelArr.sortWith(compareBy { it.m })
    var idx = 0

    for (weight in bagArr){
        while (idx < N && weight >= jewelArr[idx].m){
            priorityQueue.add(jewelArr[idx++].v)
        }
        if (priorityQueue.isNotEmpty())
            res += priorityQueue.remove()
    }
    println(res)
}

File: BOJ/Python/그리디 알고리즘/1202.py
from heapq import *
import sys
input = sys.stdin.readline
N, K = map(int, input().split())
res = 0
heap = list()
jewelArr = list()
bagArr = list()

for _ in range(N):
    jewelArr.append(tuple(map(int, input().split())))
    
for _ in range(K):
    bagArr.append(int(input()))

bagArr.sort()
jewelArr.sort(key=lambda x: x[0], reverse=True)

for weight in bagArr:
    while jewelArr and weight >= jewelArr[-1][0]:
        heappush(heap, -jewelArr.pop()[1])
    if heap:
        res -= heappop(heap)

print(res)
